adding lesson 6 (#104)
microsoft/generative-ai-with-javascript: lessons/06-tool-calling/README.md

      }
    },
    required: ["lat", "long"],
  },
  output: { type: "string" }
};

const findLandingSpotJson = {
  name: "find-landing-spot",
  description: "Finds a suitable landing spot",
  parameters: {
    type: "object",
    properties: {
      lat: {
        type: "number",
        description: "The latitude of the location",
      },
      long: {
        type: "number",
        description: "The longitude of the location",
      },
    },
    required: ["lat", "long"],
  },
  output: { type: "object", properties: { lat: "number", long: "number" } }
};

const messages = [{ 
    role: "user", 
    content: `Tell me about Amelia Earhart`,
}];

const completion = await openai.chat.completions.create({
    model: 'gpt-4o-mini',
    messages: messages,
    functions: [getBackgroundOnCharacterJson, findLandingSpotJson]
  });
```

**Time beetle**: "In the preceding code snippet we:" 

- Define the metadata for the `find-landing-spot` tool and the `get-background-on-character` tool. 
- Provide this metadata to the `client.getChatCompletions` call as part of the `functions` parameter. This tells the AI model that these tools are available to be called."

> **You**: "Got it, so the AI model will then call the right tool providing I provide a prompt that's similar to the tool's description?"

Author: ```suggestion
**You**: "Got it, so the AI model will call the appropriate if I provide a prompt that matches to the tool's description?"
```

**Time beetle**: "Almost, it will tell you what tool it thinks you should call and it will provide you with the parsed input parameters, but you need to call the tool yourself, let me me show you how."

### Calling a tool

**Time beetle**: "As I was saying, the AI model will tell you what tool it thinks you should call and it will provide you with the parsed input parameters. You then need to call the tool yourself. Here's what the workflow will look like step by step:

1. Wire up the tool call

> First, you need to wire up the tool call in your code. This involves creating the function and a metadata representation of the tool, and then providing the metadata to the AI model.

Author: ```suggestion
   First, you need to wire up the tool call in your code. This involves creating the function and a metadata representation of the tool, and then providing the metadata to the AI model.
```

}, {
    role: "user",
    content: "Find a landing spot given coordinates 8.5, 130.5"
}
];

async function main(){
  console.log("Making LLM call")

  // 6: Call the AI model with the defined messages and tools
  const result = await openai.chat.completions.create({
      model: 'gpt-4o',
      messages: messages,
      functions: [findLandingSpotJson]
    });

  for (const choice of result.choices) {

      let functionCall = choice.message?.function_call;
      let functionName = functionCall?.name;
      let args = JSON.parse(functionCall?.arguments);

      // 7: Interpret response and call the tool based on the function call provided by the AI model
      if (functionName && functionName in tools) {
          console.log(`Calling [${functionName}]`);
          const toolFunction = tools[functionName];
          const toolResponse = toolFunction(...Object.values(args)); // Extract values from args and spread them
          console.log("Result from [tool] calling: ", toolResponse);
      }
  }
}

main();
```

In the preceding code we've:

> - Create a function called `findLandingSpot` that takes latitude and longitude as input and returns the coordinates of a suitable landing spot.

Reviewer: Created ...

> - Create a function called `findLandingSpot` that takes latitude and longitude as input and returns the coordinates of a suitable landing spot.

Author: ```suggestion
- Created a function called `findLandingSpot` that takes latitude and longitude as input and returns the coordinates of a suitable landing spot.
```

# Lesson 6: Tool calling

> Tool calling, or function calling that it's also known as is about providing capabilities to your AI model it didn't have before. The idea is to provide it with meta descriptions of your actual functions and make the AI model point out when such a tool should be called given a user's prompt.

Reviewer 2: This is missing a word or two it seems. Here's another way to phrase it:

Tool calling, also known as function calling, refers to enhancing your AI model by adding capabilities it previously lacked. The concept involves supplying meta descriptions of your functions, allowing the AI model to determine when a particular tool should be called based on a user's prompt.

## Interact with Amelia Earhart

If you want to interact with Ada, run the [Characters](/app/README.md) app.

> [!IMPORTANT]
> This is entirely fictional; the responses are generated by AI.
> [Responsible AI disclaimer](/README.md#responsible-ai-disclaimer)

<div>
  <img src="./assets/amelia-front.jpeg" alt="Ada Lovelace" width="300">
</div>

**Steps**:

1. Start a [![GitHub Codespace](https://img.shields.io/badge/GitHub-Codespace-brightgreen)](https://codespaces.new/microsoft/generative-ai-with-javascript)
2. Navigate to _/app_ in the repo root.
3. Locate the console and run `npm install` followed by `npm start`.
4. Once it appears, select the "Open in Browser" button.
5. Chat with Amelia.

For a more detailed explanation of the app, see [Detailed app explanation](/lessons/01-intro-to-genai/README.md#interact-with-dinocrates).

> [!NOTE]
 > If you're running the project locally on your machine, please review the QuickStart guide to get a [GitHub personal access](/docs/setup/README.md#creating-a-personal-access-token-pat-for-github-model-access) token setup and replace the key in the code.

## Tool calling

**You**: "Device, what just happened?"

> **Time beetle**: "You just called a tool. A tool is a function that can be called by the AI model to perform a specific task. The tool can be used to perform a wide range of tasks, from simple calculations to complex operations. In this case, you called the `find-landing-spot` tool to help Amelia Earhart find a suitable landing spot."

Reviewer 2: To make it consistent with the other chapters, let's change all occurences of "Time beetle" to "Time Beetle (uppercase "B").

**Time beetle**: "Here's an image to illustrate the process of tool calling:"

<div>
    <img src="./assets/tool_call_langchain.png" alt="Tool calling process illustration" width="600">
</div>

_Image credit Langchain <https://python.langchain.com/docs/concepts/tool_calling/>_

> 

Reviewer 2: Should we add a little about error handling in here? Maybe something like this?

You: Can we handle errors if something goes wrong, like if the tool can’t find a landing spot?

Time Beetle: Great question! Yes, you can add error handling to manage such situations. For example, if the tool fails to find a landing spot, you can use a try-catch block or check the result before proceeding. Here's an example of handling errors when calling the find-landing-spot tool:

javascript
try {
    const landingSpot = findLandingSpot(7.5, 134.5);
    if (!landingSpot) {
        throw new Error("No suitable landing spot found");
    }
    console.log(`Landing spot found at coordinates: ${landingSpot.lat}, ${landingSpot.long}`);
} catch (error) {
    console.log(`Error: ${error.message}`);
}

**You**: "Got it, so the AI model will call the appropriate if I provide a prompt that matches to the tool's description?"

**Time beetle**: "Almost, it will tell you what tool it thinks you should call and it will provide you with the parsed input parameters, but you need to call the tool yourself, let me show you how."

### Calling a tool

**Time beetle**: "As I was saying, the AI model will tell you what tool it thinks you should call and it will provide you with the parsed input parameters. You then need to call the tool yourself. Here's what the workflow will look like step by step:

1. Wire up the tool call

   First, you need to wire up the tool call in your code. This involves creating the function and a metadata representation of the tool, and then providing the metadata to the AI model.

1. User makes a request via a prompt:
   - Program makes a chat completion request to the AI model with the user prompt and tools metadata provided.
   - Program receives a response from the AI model with the tool call and parsed input parameters if it thinks a tool should be called.

>    - If so, the developer interprets the response and calls the tool based on the function call provided by the AI model.

Reviewer 2: Slightly different phrasing to consider:

If so, the developer interprets the response and invokes the tool based on the function call provided by the AI model.

      }
  }
}

main();
```

In the preceding code we've:

- Created a function called `findLandingSpot` that takes latitude and longitude as input and returns the coordinates of a suitable landing spot.
- Defined the metadata for the `find-landing-spot` tool.
- Created a `tools` object that maps tool names to tool metadata.
- Provided the `tools` object to the `client.getChatCompletions` call.

   ```javascript
   if (functionName && functionName in tools) {
    console.log(`Calling [${functionName}]`);
    const toolFunction = tools[functionName];
    const toolResponse = toolFunction(...Object.values(args)); // Extract values from args and spread them
    console.log("Result from [tool] calling: ", toolResponse);
   }
   ```

- Called the tool based on the function call provided by the AI model.
- Printed the result of the tool call.

**You**: "I think I get it. I define a function, create a metadata representation of the tool, provide the metadata to the AI model, and then call the tool based on the function call provided by the AI model."

**Time beetle**: "Exactly! You're ready to start building your own tools and integrating them with the AI model."

> ## Assignment - upgrade Amelia's time travel device

Reviewer 2: "upgrade" should be "Upgrade".

**Amelia Earhart**: "We're coming down hot, thank God you found us a landing spot. Hold on tight!"

> Amelia ends up expertly landing the plane on a small island. You and Amelia step out of the plane, whereby Amelia hands you a small device.

Reviewer 2: I'd modify this sentence slightly:

Amelia expertly lands the plane on a small island.

}
```

**You**: "Time beetle, are you sure these functions are going to work, looks like they're just returning some random values?"

**Time beetle**: "That's correct, I can do the rest internally. All you need to do is to register them and test them out, make sure the AI model can call them."

> Task: Register the `calculateDistance`, `getGpsPosition`, and `getWeatherForecast` tools with the AI model. Test the tools by calling them from the AI model. Use the code supplied in the previous sections as a reference.

## Solution

[Solution](./solution/solution.js)

## Knowledge check

**Question:**  
What is the purpose of registering a tool with the AI model?

A. To allow the AI model to directly execute the tool without developer intervention.  
B. To provide metadata about the tool so the AI model can suggest its usage.  
C. To replace the need for defining functions in the code.

**Question:**  
What is the role of tool metadata in tool calling?

A. It describes the tool's purpose, inputs, and outputs for the AI model.  
B. It provides the AI model with the tool's implementation details.  
C. It ensures the tool is executed automatically by the AI model.

**Question:**  

> Why do tool calling?

Reviewer 2: Maybe change to "Why use tool calling?"

A. To enable the AI model to perform tasks beyond its built-in capabilities by leveraging external functions.  
B. To replace the need for human intervention in AI model development.  
C. To allow the AI model to execute tools without requiring metadata.

[Solution quiz](./solution/solution-quiz.md)

## Self-Study resources

- Explains the [process of tool calling](https://learn.microsoft.com/en-us/semantic-kernel/concepts/ai-services/chat-completion/function-calling/?pivots=programming-language-csharp)

> - Tool calling in the [framework Langchain](https://js.langchain.com/docs/how_to/tool_calling/)

Reviewer 2: This should be:

- Tool calling in the [LangChain.js framework](https://js.langchain.com/docs/how_to/tool_calling/)
